service-base/src/main/java/org/eclipse/hono/service/commandrouter/DelegatingCommandRouterAmqpEndpoint.java

```diff
@@ -0,0 +1,239 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.service.commandrouter;
+
+import java.net.HttpURLConnection;
+import java.time.Duration;
+import java.util.Objects;
+
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.client.ClientErrorException;
+import org.eclipse.hono.config.ServiceConfigProperties;
+import org.eclipse.hono.service.amqp.AbstractDelegatingRequestResponseEndpoint;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.CommandRouterConstants;
+import org.eclipse.hono.util.MessageHelper;
+import org.eclipse.hono.util.ResourceIdentifier;
+
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+
+/**
+ * An {@code AmqpEndpoint} for managing command router information.
+ * <p>
+ * This endpoint implements Hono's <a href="https://www.eclipse.org/hono/docs/api/command-router/">Command
+ * Router API</a>. It receives AMQP 1.0 messages representing requests and sends them to an address on the vertx
+ * event bus for processing. The outcome is then returned to the peer in a response message.
+ *
+ * @param <S> The type of service this endpoint delegates to.
+ */
+public class DelegatingCommandRouterAmqpEndpoint<S extends CommandRouterService> extends AbstractDelegatingRequestResponseEndpoint<S, ServiceConfigProperties> {
+
+    private static final String SPAN_NAME_SET_LAST_GATEWAY = "set last known gateway";
+    private static final String SPAN_NAME_REGISTER_COMMAND_CONSUMER = "register command consumer";
+    private static final String SPAN_NAME_UNREGISTER_COMMAND_CONSUMER = "unregister command consumer";
+
+    /**
+     * Creates an endpoint for a service instance.
+     *
+     * @param vertx The vert.x instance to use.
+     * @param service The service to delegate to.
+     * @throws NullPointerException if any of the parameters is {@code null};
+     */
+    public DelegatingCommandRouterAmqpEndpoint(final Vertx vertx, final S service) {
+        super(vertx, service);
+    }
+
+    @Override
+    protected Future<Message> handleRequestMessage(final Message requestMessage, final ResourceIdentifier targetAddress,
+            final SpanContext spanContext) {
+        Objects.requireNonNull(requestMessage);
```

Comment:
based on *processSetLastKnownGatewayRequest* we should also check *targetAddress* for `null` or at least handle a `null` address accordingly in *processSetLastKnownGatewayRequest* ...